### ISSUE TITLE
Cannot get past "Connector Installation" - No buttons or way to advance.

### ISSUE THREAD

**Issue by bashfulrobot** [filips123/PWAsForFirefox#509]
### Description

I am installing the extension for the first time on NixOS. I can complete the first two steps when following the wizard, but there is no obvious way to move on to the "Runtime Installation" step. I can see the binaries on my system, and I had no failures in my NixOS rebuild.

I use Home Manager to install/manage Firefox, but your Nix instructions use the regular module. So, I am wondering if that is the issue.

**Extension Page**
![image](https://github.com/filips123/PWAsForFirefox/assets/10782881/0ef9f99f-3ee1-4867-91e9-1ad6be335088)

**Binaries**
![image](https://github.com/filips123/PWAsForFirefox/assets/10782881/d4f5f636-05f3-496f-9c5c-0e5cf21ab6a1)

**Nix Config File**
![image](https://github.com/filips123/PWAsForFirefox/assets/10782881/7e45f7a3-687d-4296-b947-f9ec252dafca)

### Steps to Reproduce

1. Install the FF extension
2. Click on the extension to follow the wizard
3. Rebuild with my nix config
4. stuck at the second step

### Environment

OS: NixOS 24.11.20240524.bfb7a88 (Vicuña) x86_64
Kernel: Linux 6.6.31
Shell: bash 5.2.26
DE: Gnome 46.1
WM: Mutter (Wayland)
Mozilla Firefox 126.0
Extension Version 2.12.1
firefoxpwa 2.12.0

Thank you!

**Comment by filips123** [comment]
Yeah, home manager might be the problem. I found [this](https://github.com/NixOS/nixpkgs/pull/263404#issuecomment-2025256035) config that may work with HM, but I haven't tested it. If it doesn't work, maybe package maintainers of the NixOS package can help, as I don't have experience with Nix.

**Comment by Sikeen** [comment]
I have also not been able to get past the Connector Installation, even before starting to use HomeManager. Tried the other config as well.

**Comment by EmptyVisual** [comment]
> Yeah, home manager might be the problem. I found [this](https://github.com/NixOS/nixpkgs/pull/263404#issuecomment-2025256035) config that may work with HM, but I haven't tested it. If it doesn't work, maybe package maintainers of the NixOS package can help, as I don't have experience with Nix.

this is not specific to nixos by the way. I'm using Arch Linux with firefox installed from the main repository and I also cant get past the Connector Installation page.

**Comment by bashfulrobot** [comment]
> Yeah, home manager might be the problem. I found [this](https://github.com/NixOS/nixpkgs/pull/263404#issuecomment-2025256035) config that may work with HM, but I haven't tested it. If it doesn't work, maybe package maintainers of the NixOS package can help, as I don't have experience with Nix.

Yeah, I had a look at that link. But what that's doing is installing the extension declarative.

What I did is I just installed it through the traditional means. So it's not really applicable.  (I think). 

Plus, I saw there's another comment on this thread that this is happening on arch as well. Which removes the home manager issue from the equation regardless.

**Comment by gabenasci** [comment]
I'm having the same problem with Fedora 40, can't get past "Connector installation". I tried installing both from packagecloud and the .rpm package.

**Comment by bashfulrobot** [comment]
Seems like this is a larger issue and appears not to be platform specific. 🤔

**Comment by EmptyVisual** [comment]
> > Yeah, home manager might be the problem. I found [this](https://github.com/NixOS/nixpkgs/pull/263404#issuecomment-2025256035) config that may work with HM, but I haven't tested it. If it doesn't work, maybe package maintainers of the NixOS package can help, as I don't have experience with Nix.
> 
> this is not specific to nixos by the way. I'm using Arch Linux with firefox installed from the main repository and I also cant get past the Connector Installation page.

My issue on Arch was because I was running the Firefox application through Flatpak (so may be a different issue from the original issue posted here). There is no native host messaging set up for flatpaks currently which was why the PWA connector was not able to communicate. I swapped to Firefox through pacman through the Arch repo and it worked. Associated flatpak [xdg-desktop-portal bug](https://github.com/flatpak/xdg-desktop-portal/pull/705)

**Comment by filips123** [comment]
Yeah, if you are using Flatpak (or Snap), native messaging won't work. Otherwise, except possible problems with NixOS/HomeManager, it should work fine.

**Comment by AchyParts** [comment]
I had this issue when installing on Mac using Homebrew.

**Comment by filips123** [comment]
On macOS, you also need to link the manifest, as described in the `caveats` section. It should be something like this:

```sh
sudo mkdir -p "/Library/Application Support/Mozilla/NativeMessagingHosts"
sudo ln -sf "$HOMEBREW_PREFIX/opt/firefoxpwa/share/firefoxpwa.json" "/Library/Application Support/Mozilla/NativeMessagingHosts/firefoxpwa.json"
```

**Comment by Sikeen** [comment]
> I have also not been able to get past the Connector Installation, even before starting to use HomeManager. Tried the other config as well.

I don't use the flatpak btw

**Comment by AchyParts** [comment]
> On macOS, you also need to link the manifest, as described in the `caveats` section. It should be something like this:
> 
> ```shell
> sudo mkdir -p "/Library/Application Support/Mozilla/NativeMessagingHosts"
> sudo ln -sf "$HOMEBREW_PREFIX/opt/firefoxpwa/share/firefoxpwa.json" "/Library/Application Support/Mozilla/NativeMessagingHosts/firefoxpwa.json"
> ```

I think I accidentally separated the last line of code before, it worked now.

**Comment by filips123** [comment]
So, I think the only remaining issue here is when using NixOS. Problems with Flatpak and Snap are already known and don't need to be additionally reported, and if anyone still has problems on other systems, please create a new issue.

Now, back to the NixOS issue. Maybe @camillemndn or @pasqui23 can help, as they are the maintainers of the package and probably know Nix better than me.

**Comment by pasqui23** [comment]
Unfortunately I cannot replicate your problem

**Comment by camillemndn** [comment]
Hi, I think currently `firefoxpwa` only supports system-wide Firefox and not Home Manager. I suppose in your case the Home Manager version of Firefox, without PWAs, might take precedence over your Firefox with PWAs. Have you tried `which firefox` to see whether `firefoxpwa` appears in the Nix Store Path?

**Comment by pasqui23** [comment]
maybe try add `home.packages = [pkgs.firefoxpwa];`

**Comment by Granitosaurus** [comment]
does the extension have debug logs or something to trace what could be missing?

**Comment by Rodrigodd** [comment]
In Arch Linux, I was able to get the extension to work by copying the file from `/usr/lib/mozilla/native-messaging-hosts/firefoxpwa.json` to `~/.mozilla/native-messaging-hosts/`:

```shell
sudo cp /usr/lib/mozilla/native-messaging-hosts/firefoxpwa.json ~/.mozilla/native-messaging-hosts/
```

I did that based on the [paths described here](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/Native_manifests#linux), but not sure why the globally-visible paths weren't working.

**Comment by filips123** [comment]
For LibreWolf, you need to manually link the manifest as it uses a different location. You can check [the documentation](https://pwasforfirefox.filips.si/help/faq/#__tabbed_1_2) for instructions.

**Comment by BrettZolstick** [comment]
Leaving this here for anyone else who runs into this. If you're using home manager, use this
`programs.firefox.nativeMessagingHosts = [pkgs.firefoxpwa]`
instead of 
`programs.firefox.nativeMessagingHosts.packages = [pkgs.firefoxpwa]`
